### LinkedIn Job Postings — Initial Exploration

In [ ]:
# load main postings dataset
import pandas as pd

df = pd.read_csv("../data/postings.csv")
df.shape

(123849, 31)

In [ ]:
# preview first few rows
df.head()

,job_id,company_name,title,description,max_salary,pay_period,location,company_id,views,med_salary,...,skills_desc,listed_time,posting_domain,sponsored,work_type,currency,compensation_type,normalized_salary,zip_code,fips
0,921716,Corcoran Sawyer Smith,Marketing Coordinator,Job descriptionA leading real estate firm in N...,20.0,HOURLY,"Princeton, NJ",2774458.0,20.0,NaN,...,Requirements: \n\nWe are seeking a College or ...,1.713398e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,38480.0,8540.0,34021.0
1,1829192,NaN,Mental Health Therapist/Counselor,"At Aspen Therapy and Wellness , we are committ...",50.0,HOURLY,"Fort Collins, CO",NaN,1.0,NaN,...,NaN,1.712858e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,83200.0,80521.0,8069.0
2,10998357,The National Exemplar,Assitant Restaurant Manager,The National Exemplar is accepting application...,65000.0,YEARLY,"Cincinnati, OH",64896719.0,8.0,NaN,...,We are currently accepting resumes for FOH - A...,1.713278e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,55000.0,45202.0,39061.0
3,23221523,"Abrams Fensterman, LLP",Senior Elder Law / Trusts and Estates Associat...,Senior Associate Attorney - Elder Law / Trusts...,175000.0,YEARLY,"New Hyde Park, NY",766262.0,16.0,NaN,...,This position requires a baseline understandin...,1.712896e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,157500.0,11040.0,36059.0
4,35982263,NaN,Service Technician,Looking for HVAC service tech with experience ...,80000.0,YEARLY,"Burlington, IA",NaN,3.0,NaN,...,NaN,1.713452e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,70000.0,52601.0,19057.0


In [ ]:
# Structure and types, check columns, dtypes, non-null counts
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123849 entries, 0 to 123848
Data columns (total 31 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   job_id                      123849 non-null  int64  
 1   company_name                122130 non-null  object 
 2   title                       123849 non-null  object 
 3   description                 123842 non-null  object 
 4   max_salary                  29793 non-null   float64
 5   pay_period                  36073 non-null   object 
 6   location                    123849 non-null  object 
 7   company_id                  122132 non-null  float64
 8   views                       122160 non-null  float64
 9   med_salary                  6280 non-null    float64
 10  min_salary                  29793 non-null   float64
 11  formatted_work_type         123849 non-null  object 
 12  applies                     23320 non-null   float64
 13  original_liste

In [ ]:
# summary stats for numeric columns
df.describe()

,job_id,max_salary,company_id,views,med_salary,min_salary,applies,original_listed_time,remote_allowed,expiry,closed_time,listed_time,sponsored,normalized_salary,zip_code,fips
count,1.238490e+05,2.979300e+04,1.221320e+05,122160.000000,6280.000000,2.979300e+04,23320.000000,1.238490e+05,15246.0,1.238490e+05,1.073000e+03,1.238490e+05,123849.0,3.607300e+04,102977.000000,96434.000000
mean,3.896402e+09,9.193942e+04,1.220401e+07,14.618247,22015.619876,6.491085e+04,10.591981,1.713152e+12,1.0,1.716213e+12,1.712928e+12,1.713204e+12,0.0,2.053270e+05,50400.491887,28713.879887
std,8.404355e+07,7.011101e+05,2.554143e+07,85.903598,52255.873846,4.959738e+05,29.047395,4.848209e+08,0.0,2.321394e+09,3.622893e+08,3.989122e+08,0.0,5.097627e+06,30252.232515,16015.929825
min,9.217160e+05,1.000000e+00,1.009000e+03,1.000000,0.000000,1.000000e+00,1.000000,1.701811e+12,1.0,1.712903e+12,1.712346e+12,1.711317e+12,0.0,0.000000e+00,1001.000000,1003.000000
25%,3.894587e+09,4.828000e+01,1.435200e+04,3.000000,18.940000,3.700000e+01,1.000000,1.712863e+12,1.0,1.715481e+12,1.712670e+12,1.712886e+12,0.0,5.200000e+04,24112.000000,13121.000000
50%,3.901998e+09,8.000000e+04,2.269650e+05,4.000000,25.500000,6.000000e+04,3.000000,1.713395e+12,1.0,1.716042e+12,1.712670e+12,1.713408e+12,0.0,8.150000e+04,48059.000000,29183.000000
75%,3.904707e+09,1.400000e+05,8.047188e+06,8.000000,2510.500000,1.000000e+05,8.000000,1.713478e+12,1.0,1.716088e+12,1.713283e+12,1.713484e+12,0.0,1.250000e+05,78201.000000,42077.000000
max,3.906267e+09,1.200000e+08,1.034730e+08,9975.000000,750000.000000,8.500000e+07,967.000000,1.713573e+12,1.0,1.729125e+12,1.713562e+12,1.713573e+12,0.0,5.356000e+08,99901.000000,56045.000000


In [ ]:
# check for missing values per column
df.isnull().sum()

job_id                             0
company_name                    1719
title                              0
description                        7
max_salary                     94056
pay_period                     87776
location                           0
company_id                      1717
views                           1689
med_salary                    117569
min_salary                     94056
formatted_work_type                0
applies                       100529
original_listed_time               0
remote_allowed                108603
job_posting_url                    0
application_url                36665
application_type                   0
expiry                             0
closed_time                   122776
formatted_experience_level     29409
skills_desc                   121410
listed_time                        0
posting_domain                 39968
sponsored                          0
work_type                          0
currency                       87776
c

In [ ]:
# check for duplicates
df.duplicated().sum()

np.int64(0)

In [ ]:
# May have to merge, so we check whether each supporting table is one-to-one or one-to-many with job_id

# check how many rows per job_id in job_skills
job_skills = pd.read_csv("../data/jobs/job_skills.csv")
job_skills["job_id"].value_counts().describe()

count    126807.000000
mean          1.685774
std           0.675907
min           1.000000
25%           1.000000
50%           2.000000
75%           2.000000
max           3.000000
Name: count, dtype: float64

confirms what we suspected: job_skills.csv is one-to-many relative to job postings.

Mean of 1.69 means the average job posting has just under 2 skill entries associated with it
Max of 3 means no job has more than 3 skill entries

In [ ]:
# aggregate before merging

# check what columns job_skills actually has
job_skills.head()

,job_id,skill_abr
0,3884428798,MRKT
1,3884428798,PR
2,3884428798,WRT
3,3887473071,SALE
4,3887465684,FIN


In [ ]:
# collapse multiple skills per job into one row per job
skills_grouped = job_skills.groupby("job_id")["skill_abr"].apply(list).reset_index()
skills_grouped.head()

,job_id,skill_abr
0,921716,"[MRKT, SALE]"
1,1218575,[HCPR]
2,1829192,[HCPR]
3,2264355,"[DSGN, ART, IT]"
4,10998357,"[MGMT, MNFC]"


each job now has a single row with all its skill codes collapsed into a list

In [ ]:
# merge grouped skills into main dataframe
df_merged = df.merge(skills_grouped, on="job_id", how="left")
df_merged.shape

(123849, 32)

In [ ]:
# load skill code -> skill name lookup table
skills_lookup = pd.read_csv("../data/mappings/skills.csv")
skills_lookup.head()

,skill_abr,skill_name
0,ART,Art/Creative
1,DSGN,Design
2,ADVR,Advertising
3,PRDM,Product Management
4,DIST,Distribution


In [ ]:
# turn those skill codes into readable skill names

# build a dictionary to look up skill names from codes
skill_map = dict(zip(skills_lookup["skill_abr"], skills_lookup["skill_name"]))

# translate each job's list of skill codes into a list of skill names
df_merged["skill_names"] = df_merged["skill_abr"].apply(
    lambda codes: [skill_map.get(code, code) for code in codes] if isinstance(codes, list) else codes
)

df_merged[["job_id", "skill_abr", "skill_names"]].head()

,job_id,skill_abr,skill_names
0,921716,"[MRKT, SALE]","[Marketing, Sales]"
1,1829192,[HCPR],[Health Care Provider]
2,10998357,"[MGMT, MNFC]","[Management, Manufacturing]"
3,23221523,[OTHR],[Other]
4,35982263,[IT],[Information Technology]


In [ ]:
import pandas as pd

# list of all remaining tables to preview
tablesToCheck = {
    "companies": "../data/companies/companies.csv",
    "company_industries": "../data/companies/company_industries.csv",
    "company_specialities": "../data/companies/company_specialities.csv",
    "employee_counts": "../data/companies/employee_counts.csv",
    "benefits": "../data/jobs/benefits.csv",
    "job_industries": "../data/jobs/job_industries.csv",
    "salaries": "../data/jobs/salaries.csv",
}

for tableName, tablePath in tablesToCheck.items():
    print(f"\nTable: {tableName}")
    dfTemp = pd.read_csv(tablePath)
    print(f"Shape: {dfTemp.shape}")
    print(f"\nColumns: {dfTemp.columns.to_list()}")
    print(f"\nMissing values per column:")
    print(dfTemp.isnull().sum())
    print(f"\nFirst 3 rows:")
    print(dfTemp.head(3))
    print(f"\nDuplicate rows: {dfTemp.duplicated().sum()}")

In [ ]:
import sys

with open("../data/tablePreviews.txt", "w") as f:
    originalStdout = sys.stdout
    sys.stdout = f
    for tableName, tablePath in tablesToCheck.items():
        print(f"\nTABLE: {tableName}")
        dfTemp = pd.read_csv(tablePath)
        print(f"Shape: {dfTemp.shape}")
        print(f"\nColumns: {dfTemp.columns.tolist()}")
        print(f"\nMissing values per column:")
        print(dfTemp.isnull().sum())
        print(f"\nFirst 3 rows:")
        print(dfTemp.head(3))
        print(f"\nDuplicate rows: {dfTemp.duplicated().sum()}")
    sys.stdout = originalStdout

print("Done — check data/tablePreviews.txt")